### PR TITLE
fix(launchpad): scaffold correct config file

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes https://cypress-io.atlassian.net/browse/UNIFY-1180
- Closes https://cypress-io.atlassian.net/browse/UNIFY-925
- Closes https://cypress-io.atlassian.net/browse/UNIFY-1179
- Closes https://cypress-io.atlassian.net/browse/UNIFY-1193

### User facing changelog

- Ensure we are always recommending a `cypress.config.js` that is guaranteed to work with your framework.
- Ensure we recommend correct dependencies to install for component testing.

### Additional details

**Note**: this PR targets `lmiller1990/UNIFY-1180-add-examples`. In that branch, I add about 10 different example projects, representing the libraries/frameworks we support (CRA, Next.js, etc). I put those in a separate PR, so this PR could just contain logic. You can see all the examples here: https://github.com/cypress-io/cypress/compare/10.0-release...lmiller1990/UNIFY-1180-add-examples?expand=1

**Note**: @ZachJW34 pointed out some complexity we have with peerDependencies. That is going to be handled in a separate PR which will come after this one: https://github.com/cypress-io/cypress/pull/20277

______

Our previous approach was to try and create a config file dynamically based on your framework and bundler. There is far too many combinations for this to be reliable. In addition, there were no tests verifying the generated config file actually worked correctly, so a lot things were not working as you'd expected.

This PR does a few things:

1. move to a more simple approach of hard-coding the config for each framework/library combo. It's simple but scales, we can easily add more.
2. Improve framework detection. Now we can recommend the correct version(s) of things like `@cypress/vue`, `@cypress/webpack-dev-server` etc, depending on the version of your libraries.
3. Move all the logic around detecting frameworks out of `data-context` into a new package called `scaffold-config`.
4. Adds a project for each framework/library combo we support in `system-tests`. There's some duplication across `npm/{react,vue}`, but ultimately we want to collapse most of those tests into `system-tests`, so this is a good place to start.
5. Adds tests around the launchpad config generation workflow, at the end verifying that the generated config file matches one that's known to work (based on a system-test using the same config file).

I documented how to add more projects and configurations here: https://github.com/cypress-io/cypress/pull/20306/files#diff-8b46b4bb0349c91b80c3f521cb760e26fb444cca40a712b4dd9040384cd6bb0f

### How has the user experience changed?

Ensure we generate correct files when scaffolding your project.

**Demo** (Cypress internal team only): https://cypressio.slack.com/archives/C011BAPC614/p1645703750777719

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -65,22 +65,20 @@ npm/cypress-schematic/src/**/*.js
...
 system-tests/projects/nuxtjs-vue2-configured/**/*
-system-tests/projects/create-react-app-configured/**/*.js
-system-tests/projects/create-react-app-unconfigured/**/*.js
-system-tests/projects/vueclivue3-unconfigured/**/*
-system-tests/projects/vueclivue3-configured/**/*
-system-tests/projects/nextjs-configured/**/*
-system-tests/projects/nextjs-configured/**/*
-system-tests/projects/nuxtjs-vue2-unconfigured/**/*
-system-tests/projects/nuxtjs-vue2-configured/**/*
+
```
Comment (lmiller1990) [inline .eslintignore:83]: There are mostly community templates (think CRA, Vite, Next...) so I'd don't want to lint them - just ignoring them for now.

File: packages/data-context/src/actions/MigrationActions.ts [hunk 2]
```diff
@@ -127,4 +128,24 @@ export class MigrationActions {
       throw Error(`Expected ${actual} to equal ${expected}`)
     }
   }
+
+  async assertSuccessfulConfigScaffold (configFile: `cypress.config.${'js'|'ts'}`) {
+    assert(this.ctx.currentProject)
```
Comment (lmiller1990) [inline packages/data-context/src/actions/MigrationActions.ts:133]: This is how I am asserting the generated `cypress.config.js` is valid. It compares the newly created one to a project that is known to work and runs in CI, so if the two configs match, we can be confident it's correct.

File: packages/data-context/src/actions/WizardActions.ts [hunk 3]
```diff
@@ -28,28 +29,33 @@ export class WizardActions {
...
+
+    const prevFramework = this.ctx.coreData.wizard.chosenFramework || ''
 
-    return
+    if (doesNotSupportChosenBundler || !['react', 'vue'].includes(prevFramework)) {
+      this.setBundler(null)
+    }
   }
 
-  setBundler (bundler: NexusGenEnums['SupportedBundlers'] | null) {
```
Comment (lmiller1990) [inline packages/data-context/src/actions/WizardActions.ts:52]: I don't think we should use nexus codegen types in `data-context`... rather we should use the ones we define ourselves. I would like to lint for this in the future.

File: packages/data-context/src/actions/WizardActions.ts [hunk 4]
```diff
@@ -84,75 +90,33 @@ export class WizardActions {
...
-      debug('detectedFramework %s', this.data.detectedFramework)
-      this.detectBundler(dependencies)
-      debug('detectedBundler %s', this.data.detectedBundler)
-
-      this.data.chosenFramework = this.data.detectedFramework || null
-      this.data.chosenBundler = this.data.detectedBundler || null
-    }
-  }
-
-  private detectFramework (dependencies: string[]) {
```
Comment (lmiller1990) [inline packages/data-context/src/actions/WizardActions.ts:124]: All this detection stuff moved to `@packages/scaffold-config` and is covered head-to-toe with tests.

File: packages/data-context/src/actions/WizardActions.ts [hunk 5]
```diff
@@ -315,28 +274,6 @@ export class WizardActions {
     return codeBlocks.join('\n')
   }
 
-  private wizardGetConfigCodeComponent (opts: WizardGetCodeComponent): string {
```
Comment (lmiller1990) [inline packages/data-context/src/actions/WizardActions.ts:318]: This all moved to `@packages/scaffold-config`

File: packages/data-context/test/unit/sources/WizardDataSource.spec.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
+import { expect } from 'chai'
+import path from 'path'
+import { e2eProjectDirs } from '@packages/frontend-shared/cypress/e2e/support/e2eProjectDirs'
+import { createTestDataContext } from '../helper'
+
+function getCurrentProject (project: typeof e2eProjectDirs[number]) {
+  return path.join(__dirname, '..', '..', '..', '..', '..', 'system-tests', 'projects', project)
+}
+
+describe('packagesToInstall', () => {
```
Comment (lmiller1990) [inline packages/data-context/test/unit/sources/WizardDataSource.spec.ts:10]: The tests we so desperately need

File: packages/scaffold-config/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+/* eslint-disable padding-line-between-statements */
```
Comment (lmiller1990) [inline packages/scaffold-config/src/index.ts:1]: The file above this is "Load diff" but you should look at it, that's where all the frameworks and their configuration is.

File: packages/scaffold-config/src/frameworks.ts [hunk 8]
```diff
@@ -0,0 +1,579 @@
...
+  WEBPACK_DEV_SERVER_4,
+  BUNDLER_WEBPACK_4,
+  CYPRESS_WEBPACK,
+} from './constants'
+import { BUNDLER_VITE, BUNDLER_WEBPACK_5, HTML_WEBPACK_PLUGIN_5 } from '.'
+
+const isWebpack = (bundler: Bundler) => ['webpack4', 'webpack5'].includes(bundler)
+
+export const FRONTEND_FRAMEWORKS = [
+  {
```
Comment (elevatebart) [inline packages/scaffold-config/src/frameworks.ts:21]: Could each of those be in its own file?
I could see myself scrolling for hours to find the framework I am looking to debug

File: packages/launchpad/cypress/e2e/project-setup.cy.ts [hunk 9]
```diff
@@ -1,4 +1,15 @@
-import { FRONTEND_FRAMEWORKS, BUNDLERS, CODE_LANGUAGES, PACKAGES_DESCRIPTIONS } from '@packages/types/src/constants'
+import { BUNDLERS, FRONTEND_FRAMEWORKS, AllPackagePackages } from '@packages/scaffold-config/src'
+import { CODE_LANGUAGES } from '@packages/types/src'
+
+function fakeInstalledDeps () {
```
Comment (elevatebart) [inline packages/launchpad/cypress/e2e/project-setup.cy.ts:4]: I like that 😍 TYTYTY

**Review by elevatebart — APPROVED** [review]
Outstanding job overall! Thank you for taking this on.

A few small remarks

File: packages/scaffold-config/src/detect.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    ],
+  },
+] as const
+
+// Detect the framework, which can either be a tool like Create React App,
+// in which case we just return the framework. The user cannot change the
+// bundler.
+
+// If we don't find a specific framework, but we do find a library and/or
+// bunlder, we return both the framework, which might just be "React",
```
[the remark this answers is not in the document]
Comment (elevatebart) [reply packages/scaffold-config/src/detect.ts:50]:
```suggestion
// bundler, we return both the framework, which might just be "React",
```

File: packages/frontend-shared/cypress/support/mock-graphql/stubgql-Wizard.ts [hunk 11]
```diff
@@ -14,16 +15,17 @@ export const allBundlers = BUNDLERS.map((bundler, idx) => {
...
 export const stubWizard: MaybeResolver<Wizard> = {
   __typename: 'Wizard',
+  installDependenciesCommand: 'npm install -D @cypress/react @cypress/webpack-dev-server',
   packagesToInstall: [
     {
       ...testNodeId('WizardNpmPackage'),
-      description: PACKAGES_DESCRIPTIONS['@cypress/react'],
+      description: CYPRESS_REACT_LATEST.description,
       name: '@cypress/react',
       package: '@cypress/react',
```
Comment (elevatebart) [inline packages/frontend-shared/cypress/support/mock-graphql/stubgql-Wizard.ts:24]:
```suggestion
      ... CYPRESS_REACT_LATEST,
```

File: packages/frontend-shared/cypress/support/mock-graphql/stubgql-Wizard.ts [hunk 12]
```diff
@@ -14,16 +15,17 @@ export const allBundlers = BUNDLERS.map((bundler, idx) => {
...
     },
     {
       ...testNodeId('WizardNpmPackage'),
-      description: PACKAGES_DESCRIPTIONS['@cypress/webpack-dev-server'],
+      description: CYPRESS_WEBPACK.description,
       name: '@cypress/webpack-dev-server',
       package: '@cypress/webpack-dev-server',
```
Comment (elevatebart) [inline packages/frontend-shared/cypress/support/mock-graphql/stubgql-Wizard.ts:30]:
```suggestion
      ...CYPRESS_WEBPACK.description,
```

File: packages/scaffold-config/src/frameworks.ts [hunk 13]
```diff
@@ -0,0 +1,579 @@
...
+      })`
+      },
+    },
+  },
+  {
+    type: 'vueclivue2',
+    name: 'Vue CLI (Vue 2)',
+    family: 'template',
+    supportedBundlers: [BUNDLER_WEBPACK_4],
+    packages: [CYPRESS_VUE_2, CYPRESS_WEBPACK],
```
Comment (ZachJW34) [inline packages/scaffold-config/src/frameworks.ts:120]: Should we do the same thing here that we do with `cra` i.e. list all webpack deps?

File: packages/scaffold-config/src/frameworks.ts [hunk 14]
```diff
@@ -0,0 +1,579 @@
...
+          devServerConfig: {
+            webpackConfig
+          }
+        }
+      })`
+      },
+    },
+  },
+  {
+    type: 'vueclivue3',
```
Comment (ZachJW34) [inline packages/scaffold-config/src/frameworks.ts:170]: Vue-cli v5 is stable which uses webpack 5, fairly new but I'd consider adding a variant or follow up soon after.

File: packages/scaffold-config/src/frameworks.ts [hunk 15]
```diff
@@ -0,0 +1,579 @@
...
+    detectors: [
+      {
+        dependency: 'next',
+        version: '>=10.0.0',
+      },
+    ],
+    codeGenFramework: CODE_GEN_FRAMEWORKS[0],
+    storybookDep: STORYBOOK_REACT,
+    config: {
+      js: () => ``,
```
Comment (ZachJW34) [inline packages/scaffold-config/src/frameworks.ts:518]: I can help out with Next if need be.

File: packages/data-context/src/actions/WizardActions.ts [hunk 16]
```diff
@@ -84,75 +90,33 @@ export class WizardActions {
...
+        coreData.wizard.detectedBundler = null
+        coreData.wizard.detectedLanguage = null
+      })
 
       await this.detectLanguage()
       debug('detectedLanguage %s', this.data.detectedLanguage)
       this.data.chosenLanguage = this.data.detectedLanguage || 'js'
 
-      let hasPackageJson = true
+      const detected = detect(await fs.readJson(path.join(this.ctx.currentProject, 'package.json')))
```
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/data-context/src/actions/WizardActions.ts:103]: 0f9206978fb093d2d1905a3c148b066d818b6035

File: packages/frontend-shared/cypress/support/mock-graphql/stubgql-Wizard.ts [hunk 17]
```diff
@@ -14,16 +15,17 @@ export const allBundlers = BUNDLERS.map((bundler, idx) => {
 
 export const stubWizard: MaybeResolver<Wizard> = {
   __typename: 'Wizard',
+  installDependenciesCommand: 'npm install -D @cypress/react @cypress/webpack-dev-server',
   packagesToInstall: [
     {
       ...testNodeId('WizardNpmPackage'),
-      description: PACKAGES_DESCRIPTIONS['@cypress/react'],
+      description: CYPRESS_REACT_LATEST.description,
```
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/frontend-shared/cypress/support/mock-graphql/stubgql-Wizard.ts:22]: eb694fe85fcce4869171fa41a87c7d59fbab605a

File: packages/frontend-shared/cypress/support/mock-graphql/stubgql-Wizard.ts (hunk 12 above)
Comment (lmiller1990) [reply packages/frontend-shared/cypress/support/mock-graphql/stubgql-Wizard.ts:30]: eb694fe85fcce4869171fa41a87c7d59fbab605a

File: packages/scaffold-config/src/detect.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/scaffold-config/src/detect.ts:50]: c64eda4a6c22052393f2b18c5e4d5bd4868ed5bc

File: packages/scaffold-config/src/frameworks.ts (hunk 13 above)
Comment (lmiller1990) [reply packages/scaffold-config/src/frameworks.ts:120]: Done for `template` 107cd32209e574d683e53697ffab95035add9a35

File: packages/scaffold-config/src/frameworks.ts (hunk 14 above)
Comment (lmiller1990) [reply packages/scaffold-config/src/frameworks.ts:170]: For sure, maybe you can follow these instructions to add it? https://github.com/cypress-io/cypress/pull/20350/files#diff-8b46b4bb0349c91b80c3f521cb760e26fb444cca40a712b4dd9040384cd6bb0f and update them with anything that's missing so other people can easily add new projects in the future.

File: packages/scaffold-config/src/frameworks.ts (hunk 15 above)
Comment (lmiller1990) [reply packages/scaffold-config/src/frameworks.ts:518]: :pray:

File: packages/scaffold-config/test/unit/detect.spec.ts [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+
+export async function scaffoldMigrationProject (project: typeof e2eProjectDirs[number]) {
+  Fixtures.removeProject(project)
+
+  await Fixtures.scaffoldProject(project)
+
+  return fs.readJSONSync(path.join(Fixtures.projectPath(project), 'package.json'))
+}
+
+describe('detect', () => {
```
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/scaffold-config/test/unit/detect.spec.ts:16]: - We have react-webpack: https://github.com/cypress-io/cypress/pull/20350/files#diff-2792820e59ebe62562090f1a6a594d7c94f099670780bad69edadbff1e98ff88R31
- vue3-vite: d26dfab464e5ef3ee963194d2bdf1baf11d99b0

Follow up PR for the rest, maybe we can add w/ the next.js PR?

File: packages/scaffold-config/src/frameworks.ts [hunk 19]
```diff
@@ -0,0 +1,579 @@
...
+      })`
+      },
+    },
+  },
+  {
+    type: 'react',
+    name: 'React.js',
+    family: 'library',
+    supportedBundlers: [BUNDLER_WEBPACK_4, BUNDLER_WEBPACK_5, BUNDLER_VITE],
+    packages: [CYPRESS_REACT_LATEST],
```
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/scaffold-config/src/frameworks.ts:225]: check this out... https://github.com/cypress-io/cypress/pull/20350/commits/86081dcba5a3effc89db2814027ad863bfdaaf53#diff-81c72c844fb495ec6c551642258c9a561569dc654bbcc62373a59ebf5d730199R106-R109

Now we just declare the deps in the actual bundler, way better!

**Review by davidmunechika — APPROVED** [review]

**Review by ZachJW34 — APPROVED** [review]
Tested and tested, works great!

**Comment by elevatebart** [comment]
Hello @lmiller1990 

Why is the base branch on `...add-examples` instead of on 10.0-release?
Is it to facilitate review?
Is there a PR for this other branch that we are not aware of?

**Comment by ZachJW34** [comment]
@elevatebart The target branch has a bunch of system tests projects that would have bloated the diff so he kept the logic separate

**Pull request merged**